### PR TITLE
Workspace token switch support

### PR DESCRIPTION
**Pull request by chandumlg**
Workspace token switch support
Backend-config: removed default workspace token
Moved postgres compatibility check to envValidator
Logger: warn level introduced

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/180)

### CODE REVIEW

File: config/backend-config/backend-config.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func loadConfig() {
 	multiWorkspaceSecret = config.GetEnv("HOSTED_SERVICE_SECRET", "password")
 
 	configBackendURL = config.GetEnv("CONFIG_BACKEND_URL", "https://api.rudderlabs.com")
-	configBackendToken = config.GetEnv("CONFIG_BACKEND_TOKEN", "1P2tfQQKarhlsG6S3JGLdXptyZY")
+	configBackendToken = config.GetEnv("CONFIG_BACKEND_TOKEN", "")
```
[the remark this answers is not in the document]
Comment (chandumlg) [reply config/backend-config/backend-config.go]: Done

File: services/validators/envValidator.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	var versionNum int
+	err := dbHandle.QueryRow("SHOW server_version_num;").Scan(&versionNum)
+	if err != nil {
+		return false
+	}
+
+	return versionNum >= minPostgresVersion
+}
+
+func didWorkspaceTokenChange() bool {
```
[the remark this answers is not in the document]
Comment (chandumlg) [reply services/validators/envValidator.go]: Done

File: services/validators/envValidator.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	if totalCount > 0 {
+		return
+	}
+
+	//There are no entries in the table, insert current workspace token
+	insertTokenIntoWorkspace(config.GetEnv("CONFIG_BACKEND_TOKEN", ""))
```
[the remark this answers is not in the document]
Comment (chandumlg) [reply services/validators/envValidator.go]: Done

File: services/validators/envValidator.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		return true
+	}
+
+	//db connection should be closed. Else alter db fails.
+	dbHandle.Close()
+
+	logger.Warn("Previous workspace token is not same as the current workspace token. Parking current jobsdb aside and creating a new one")
+
+	dbName := config.GetEnv("JOBS_DB_DB_NAME", "ubuntu")
+	misc.ReplaceDB(dbName, dbName+"_"+strconv.FormatInt(time.Now().Unix(), 10)+"_"+config.GetEnv("CONFIG_BACKEND_TOKEN", ""))
```
[the remark this answers is not in the document]
Comment (chandumlg) [reply services/validators/envValidator.go]: Done

File: utils/misc/misc.go [hunk 5]
```diff
@@ -122,6 +124,24 @@ func AssertErrorIfDev(err error) {
 	}
 }
 
+//GetWorkspaceToken returns the workspace token provided in the *.env file.
```
Comment (SumanthPuram) [inline utils/misc/misc.go]: Can we move this to config.go?

File: services/validators/envValidator.go [hunk 6]
```diff
@@ -127,12 +125,14 @@ func ValidateEnv() bool {
 	logger.Warn("Previous workspace token is not same as the current workspace token. Parking current jobsdb aside and creating a new one")
 
 	dbName := config.GetEnv("JOBS_DB_DB_NAME", "ubuntu")
-	misc.ReplaceDB(dbName, dbName+"_"+strconv.FormatInt(time.Now().Unix(), 10)+"_"+config.GetEnv("CONFIG_BACKEND_TOKEN", ""))
+	misc.ReplaceDB(dbName, dbName+"_"+strconv.FormatInt(time.Now().Unix(), 10)+"_"+workspaceTokenHashInDB[:32])
```
Comment (SumanthPuram) [inline services/validators/envValidator.go]: Why are we picking only 32 chars?

**Pull request merged**